data-model/src/main/java/org/hypertrace/core/datamodel/shared/StructuredTraceGraph.java

```diff
@@ -20,187 +14,81 @@
  */
 public class StructuredTraceGraph {
 
-  /* there could be multiple roots for partial trace and incomplete instrumented app spans */
-  private final Map<ByteBuffer, List<Event>> parentToChildrenEvents;
-  private final Map<ByteBuffer, Event> childToParentEvents;
-  /* entity have many-to-many relationship */
-  private final Map<String, List<Entity>> parentToChildrenEntities;
-  private final Map<String, List<Entity>> childToParentEntities;
-
-  /* these containers should be unmodifiable after initialization as we're exposing them via getters */
-  private Map<ByteBuffer, Event> eventMap;
-  private Map<String, Entity> entityMap;
-  private Set<Event> rootEvents;
-  private Set<Entity> rootEntities;
-
-  private StructuredTraceGraph() {
-    this.childToParentEntities = new HashMap<>();
-    this.parentToChildrenEntities = new HashMap<>();
-    this.childToParentEvents = new HashMap<>();
-    this.parentToChildrenEvents = new HashMap<>();
-  }
+  private TraceEventsGraph traceEventsGraph;
+  private TraceEntitiesGraph traceEntitiesGraph;
+
+  private static StructuredTraceGraph graph;
 
-  /**
-   * Create the instance with or without the full traversal graph includesFullGraph    Includes
-   * computing the full traversal graph or not
-   */
   public static StructuredTraceGraph createGraph(StructuredTrace trace) {
-    StructuredTraceGraph graph = new StructuredTraceGraph();
-    graph.buildEventMap(trace);
-    graph.buildEntityMap(trace);
-    graph.buildParentChildRelationship(trace);
-    graph.buildRootEvents(trace);
-    graph.buildRootEntities(trace);
+    graph = new StructuredTraceGraph();
+    graph.traceEventsGraph = TraceEventsGraph.createGraph(trace);
+    graph.traceEntitiesGraph = TraceEntitiesGraph.createGraph(trace);
+
+    return graph;
+  }
+
+  public static StructuredTraceGraph reCreateTraceEventsGraph(StructuredTrace trace) {
+    if (null == graph) {
+      return createGraph(trace);
+    }
+    graph.traceEventsGraph = TraceEventsGraph.createGraph(trace);
```

Comment:
Case where graph is null, we seem to be calling createGraph which internally does `TraceEventsGraph.createGraph(trace)`. So this seems like a duplicate call.

This holds true for all the recreate methods.